pkg/kubelet/sysctl/safe_sysctls.go

```diff
@@ -51,19 +51,31 @@ var safeSysctlsIncludeReservedPorts = []string{
 // - it is namespaced in the container or the pod
 // - it is isolated, i.e. has no influence on any other pod on the same node.
 func SafeSysctlAllowlist() []string {
-	if goruntime.GOOS == "linux" {
-		// make sure we're on a new enough kernel that the ip_local_reserved_ports sysctl is namespaced
-		kernelVersion, err := getKernelVersion()
-		if err != nil {
-			klog.ErrorS(err, "Failed to get kernel version, dropping net.ipv4.ip_local_reserved_ports from safe sysctl list")
-			return safeSysctls
-		}
-		if kernelVersion.LessThan(version.MustParseGeneric(ipLocalReservedPortsMinNamespacedKernelVersion)) {
-			klog.ErrorS(nil, "Kernel version is too old, dropping net.ipv4.ip_local_reserved_ports from safe sysctl list", "kernelVersion", kernelVersion)
-			return safeSysctls
-		}
+	if goruntime.GOOS != "linux" {
+		return nil
 	}
-	return safeSysctlsIncludeReservedPorts
+
+	kernelVersion, err := getKernelVersion()
+	if err != nil {
+		klog.ErrorS(err, "failed to get kernel version, unable to determine which sysctls are available")
+		return safeSysctlsNoLimitedByKernelVersion
+	}
+
+	var safeSysctls []string
```

Comment:
This is OK for one or two, but will be a mess much beyond that.  Why not retool this like:

```
type sysctl struct {
    name string    // e.g. "foo.bar.qux_timeout"
    kernel string  // "" means any kernel will satisfy
}

var safeSysctls []sysctl{{
        name: "kernel.shm_rmid_forced",
    }, {
        name: "net.ipv4.ip_local_port_range",
    }, {
        name: "net.ipv4.tcp_syncookies",
    }, {
        name: "net.ipv4.ping_group_range",
    }, {
        name: "net.ipv4.ip_unprivileged_port_start",
    }, {
        name: "net.ipv4.ip_local_reserved_ports",
        kernel: "3.16",
    }, {
        name: "net.ipv4.tcp_keepalive_time",
        kernel: "4.5",
    }}
```

Then iterate that list, checking for `kernel == ""`.

Reply:
Kernel versions need to be better identified and understood in kube now, especially in kubelet.
- https://github.com/kubernetes/kubernetes/issues/116799

Reply:
Thanks, updated.